onnxruntime/core/framework/tensorprotoutils.cc

```diff
@@ -726,9 +748,30 @@ Status TensorProtoToMLValue(const Env& env, const ORTCHAR_T* model_path,
   // Note: We permit an empty tensor_shape_vec, and treat it as a scalar (a tensor of size 1).
   TensorShape tensor_shape{tensor_shape_vec};
 
+  std::vector<std::function<void(void)>> tensor_deleters;
+  ResourceGuard exitguard([&tensor_deleters]() {
+    for (const auto& d : tensor_deleters) {
+      d();
+    }
+  });
+
+  if (m.IsOwner()) {
+    // need to transfer buffer ownership to tensor
+    if (deleter.f != nullptr) {
+      // clean up should happen eariler than releasing buffer
+      OrtCallback tcall;
+      MoveOrtCallback(deleter, tcall);
+      tensor_deleters.push_back([tcall]() { tcall.f(tcall.param); });
+    }
+    ORT_ENFORCE(deleter.f == nullptr);
+    tensor_deleters.push_back(CreateBufDelClr(m.Release(), m.GetBuffer()));
```

Comment:
ORT_ENFORCE costs binary size in release builds, so it would be better to do this basic correctness testing via unit tests.